lib/rubygems/installer.rb

```diff
@@ -474,6 +474,26 @@ def generate_windows_script(filename, bindir)
     end
   end
 
+  ##
+  # Checks if +bin_path+ should be wrapped if we're generating wrappers
+  #
+  # If the first line of the file is a shebang check if it mentions ruby and
+  # return false if it doesn't (don't wrap non-ruby executable)
+  #
+  # If the first line is not a shebang, check if it contains the magic string:
+  # "rubygems: no-wrap" and return false if it does (don't wrap if no-wrap was
+  # requested)
+  #
+  # Otherwise wrap
+
+  def wrappable_executable?(bin_path)
+    File.open bin_path, 'r' do |f|
+      line1 = f.gets
+      break line1['ruby'] if line1.start_with?('#!') # if the line is a shebang check for ruby
+      !line1['rubygems: no-wrap'] # check for magic string
+    end
```

Comment:
Using String#[] is not the best way to check if a string contains a substring, as it's not a boolean operation.
Also, the method should be aware that IO#gets may return nil if the file is empty.

You could use a regexp here.

```ruby
# If the first line is a non-ruby shebang or contains a magic string, return false.
/\A#!(?!.*ruby)|rubygems: no-wrap/ !~ File.open(bin_path, &:gets)
```